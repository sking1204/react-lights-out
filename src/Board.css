.Board {
  margin: 0 auto;
  border: 2px solid black;
  border-collapse: collapse;
}