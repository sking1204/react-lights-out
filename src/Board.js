import React, { useState } from "react";
import Cell from "./Cell";
import "./Board.css";

/** Game board of Lights out.
 *
 * Properties:
 *
 * - nrows: number of rows of board
 * - ncols: number of cols of board
 * - chanceLightStartsOn: float, chance any cell is lit at start of game
 *
 * State:
 *
 * - board: array-of-arrays of true/false
 *
 *    For this board:
 *       .  .  .
 *       O  O  .     (where . is off, and O is on)
 *       .  .  .
 *
 *    This would be: [[f, f, f], [t, t, f], [f, f, f]]
 *
 *  This should render an HTML table of individual <Cell /> components.
 *
 *  This doesn't handle any clicks --- clicks are on individual cells
 *
 **/

function Board({ nrows =4, ncols=4, chanceLightStartsOn=.9}) {
  const [board, setBoard] = useState([]);
  const [gameStarted, setGameStarted] = useState(false);

  /** create a board nrows high/ncols wide, each cell randomly lit or unlit */
  function createBoard() {
    let initialBoard = [];
    // TODO: create array-of-arrays of true/false values
    for (let i = 0; i < nrows; i++) {
      let row = [];
      for (let j = 0; j < ncols; j++) {
        // Assign true/false randomly based on chanceLightStartsOn
        row.push(Math.random() < chanceLightStartsOn);
        // row.push(true)
        
      }
      initialBoard.push(row);       
      console.log("initial board:", initialBoard)
      
    }
   
    return initialBoard;
  }

  function hasWon() {
    // TODO: check the board in state to determine whether the player has won.
    return board.every(row => row.every(cell => !cell));
  }

  function flipCellsAround(coord) {
    setBoard(oldBoard => {
      const [y, x] = coord.split("-").map(Number);

      const flipCell = (y, x, boardCopy) => {
        // if this coord is actually on board, flip it

        if (x >= 0 && x < ncols && y >= 0 && y < nrows) {
          boardCopy[y][x] = !boardCopy[y][x];
        }
      };

      // TODO: Make a (deep) copy of the oldBoard
      const boardCopy = oldBoard.map(row => [...row]);

      // TODO: in the copy, flip this cell and the cells around it
      flipCell(y, x, boardCopy);
      flipCell(y, x - 1, boardCopy);
      flipCell(y, x + 1, boardCopy);
      flipCell(y - 1, x, boardCopy);
      flipCell(y + 1, x, boardCopy);

      // TODO: return the copy
      return boardCopy;
    });
  }

//added function to start game:
  function startGame(){
    setBoard(createBoard());
    setGameStarted(true);

  }

//adding logic to check if game has started:

if (!gameStarted){
  return <div><button onClick={startGame}>Start Game</button></div>;
}

  // if the game is won, just show a winning msg & render nothing else

  if (hasWon()) {
    return <div>You Win!
      <div>
      <button onClick={startGame}>Restart Game!</button> 
      </div>
     
    </div>;
  }

  // TODO

  // make table board

  let tblBoard = [];

  for (let y = 0; y < nrows; y++) {
    let row = [];
    for (let x = 0; x < ncols; x++) {
      let coord = `${y}-${x}`;
      row.push(
          <Cell
              key={coord}
              isLit={board[y][x]}
              flipCellsAroundMe={evt => flipCellsAround(coord)}
          />,
      );
    }
    tblBoard.push(<tr key={y}>{row}</tr>);
  }

  return (            
      <table className="Board">
        <tbody>{tblBoard}</tbody>
      </table>    
      
  );
}

  // TODO


export default Board;
