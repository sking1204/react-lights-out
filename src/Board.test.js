import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Board from "./Board";

describe("<Board /> rendering", function () {
  describe("inital board state", function () {
    it("renders without crashing", function () {
      render(<Board />);
    });

    it("matches snapshot for full board", function () {
      const { asFragment } = render(<Board chanceLightStartsOn={1} />);
      expect(asFragment()).toMatchSnapshot();
    });

    it("shows initial board state for new game", function(){
        const {getByText} = render (<Board chanceLightStartsOn = {0} />)
        expect(getByText("Start Game")).toBeInTheDocument();
    })

  });

  //NEED TO WORK ON CELL CLICK TEST
});
