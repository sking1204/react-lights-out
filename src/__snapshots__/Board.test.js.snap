// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Board /> rendering inital board and win state matches snapshot for full board 1`] = `
<DocumentFragment>
  <div>
    <button>
      Start Game
    </button>
  </div>
</DocumentFragment>
`;

exports[`<Board /> rendering inital board state matches snapshot for full board 1`] = `
<DocumentFragment>
  <div>
    <button>
      Start Game
    </button>
  </div>
</DocumentFragment>
`;
